packages/Core/iterable.candy

```diff
@@ -0,0 +1,167 @@
+    previousState | next | result.map { itemAndRest ->
+      item = itemAndRest | list.get 0
+      rest = itemAndRest | list.get 1
+      (mapper item, rest)
+    }
+  }
+
+withIndex iterable :=
+  needs (is iterable)
+  (iterable, 0) | rawMapWithState { previousState ->
+    iterable = previousState | list.get 0
+    index = previousState | list.get 1
+    iterable | next | result.map { itemAndRest ->
+      item = itemAndRest | list.get 0
+      rest = itemAndRest | list.get 1
+      ((item, index), (rest, index | int.add 1))
+    }
+  }
+
+## Reduction
+
+# TODO: ensure this works with tail recursion
+foldLeftHelper = { foldLeftHelper iterable value combine ->
+  iterable
+  | next
+  | result.map { it ->
+    currentItem = it | list.get 0
+    remainingIterable = it | list.get 1
+    newValue = combine value currentItem
+    foldLeftHelper foldLeftHelper remainingIterable newValue combine
+  }
+  | result.unwrapOr value
+}
+foldLeft iterable initialValue combine :=
+  needs (is iterable)
+  needs (function.is2 combine)
+  foldLeftHelper foldLeftHelper iterable initialValue combine
+# TODO: `foldRight`
+
+reduceLeft iterable combine :=
+  needs (is iterable)
+  needs (function.is2 combine)
+  iterable
+  | next
+  | result.mapError { _ -> "Can't `reduceLeft` an empty iterable" }
+  | result.map { it ->
+    currentItem = it | list.get 0
+    remainingIterable = it | list.get 1
+    remainingIterable | foldLeft currentItem combine
+  }
+# TODO: `reduceRight`
+
+sum iterableOfInt :=
+  needs (is iterableOfInt)
+  iterableOfInt | reduceLeft { a b ->
+    needs (int.is a)
+    needs (int.is b)
+    a | int.add b
+  }
+length iterable :=
+  needs (is iterable)
+  iterable | map { _ -> 1 } | sum | result.unwrapOr 0
+
+all iterable tester :=
+  needs (is iterable)
+  needs (function.is1 tester)
+  iterable | foldLeft True { result item ->
+    itemResult = tester item
+    needs (bool.is itemResult)
+    result | bool.and itemResult
+  }
+  # TODO: stop after the first `False`
+any iterable tester :=
+  needs (is iterable)
+  needs (function.is1 tester)
+  iterable | foldLeft False { result item ->
+    itemResult = tester item
+    needs (bool.is itemResult)
+    result | bool.or itemResult
+  }
+  # TODO: stop after the first `True`
+
+## Filtering
+
+# whereWithStateHelper = { whereWithStateHelper iterable state tester ->
+#   iterable | result.map { it ->
+#     currentItem = it | list.get 0
+#     remainingIterable = it | list.get 1
+#     (
+#       mapper currentItem,
+#       mapHelper mapHelper remainingIterable mapper,
+#     )
+#   }
+# }
+# whereWithState iterable initialState tester :=
+#   needs (is iterable)
+#   needs (function.is2 tester)
+#   whereWithStateHelper whereWithStateHelper iterable initialState tester
+# where iterable tester :=
+#   needs (is iterable)
+#   needs (function.is1 tester)
+
+## Action
+
+forEach iterable action :=
+  needs (is iterable)
+  needs (function.is1 action)
+  iterable | map action
```

Comment:
This won't behave as expected because `map` is lazy, while `forEach` shouldn't be. For example, `("Foo", "Bar") | listToIterable | forEach { item -> print item }` would just return a closure rather than printing the items.

Maybe change it into this?

```suggestion
forEach iterable action :=
  needs (is iterable)
  needs (function.is1 action)
  iterable
  | map { item ->
    action item
    Nothing
  }
  | toList
```